### PR TITLE
Add OnPrepareResponseAsync

### PR DESCRIPTION
**Pull request by stefannikolei**
Add OnPrepareResponseAsync to StaticFileMiddleware #44932

### CODE REVIEW

File: src/Middleware/StaticFiles/src/StaticFileContext.cs [hunk 1]
```diff
@@ -287,12 +290,11 @@ private static PreconditionState GetMaxPreconditionState(params PreconditionStat
         return max;
     }
 
-    public Task SendStatusAsync(int statusCode)
+    public async Task SendStatusAsync(int statusCode)
     {
-        ApplyResponseHeaders(statusCode);
+         await ApplyResponseHeadersAsync(statusCode);
```
Comment (JamesNK) [inline src/Middleware/StaticFiles/src/StaticFileContext.cs:295]:
```suggestion
        await ApplyResponseHeadersAsync(statusCode);
```

File: src/Middleware/StaticFiles/src/StaticFileOptions.cs [hunk 2]
```diff
@@ -62,4 +64,10 @@ public StaticFileOptions(SharedOptions sharedOptions) : base(sharedOptions)
     /// This can be used to add or change the response headers.
     /// </summary>
     public Action<StaticFileResponseContext> OnPrepareResponse { get; set; }
```
Comment (JamesNK) [inline src/Middleware/StaticFiles/src/StaticFileOptions.cs:66]:
```suggestion
    /// </summary>
    /// <remarks>
    /// <see cref="OnPrepareResponse" /> is called before <see cref="OnPrepareResponseAsync" />.
    /// </remarks>
    public Action<StaticFileResponseContext> OnPrepareResponse { get; set; }
```

File: src/Middleware/StaticFiles/src/StaticFileOptions.cs [hunk 3]
```diff
@@ -62,4 +64,10 @@ public StaticFileOptions(SharedOptions sharedOptions) : base(sharedOptions)
...
+
+    /// <summary>
+    /// Called after the status code and headers have been set, but before the body has been written.
+    /// This can be used to add or change the response headers.
+    /// </summary>
+    public Func<StaticFileResponseContext, Task> OnPrepareResponseAsync { get; set; }
```
Comment (JamesNK) [inline src/Middleware/StaticFiles/src/StaticFileOptions.cs:72]:
```suggestion
    /// </summary>
    /// <remarks>
    /// <see cref="OnPrepareResponse" /> is called before <see cref="OnPrepareResponseAsync" />.
    /// </remarks>
    public Func<StaticFileResponseContext, Task> OnPrepareResponseAsync { get; set; }
```

**Review by Tratcher — COMMENTED** [review]
Please add a unit test for each callback, and their ordering. This should be a good example:
https://github.com/dotnet/aspnetcore/blob/d0f60f2b65c09d95a794b0f8aa8d22dc78561f2d/src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs#L175

File: src/Middleware/StaticFiles/src/StaticFileContext.cs [hunk 4]
```diff
@@ -289,9 +292,8 @@ private static PreconditionState GetMaxPreconditionState(params PreconditionStat
 
     public Task SendStatusAsync(int statusCode)
     {
-        ApplyResponseHeaders(statusCode);
+        ApplyResponseHeadersAsync(statusCode);
 
-        _logger.Handled(statusCode, SubPath);
         return Task.CompletedTask;
```
Comment (campersau) [inline src/Middleware/StaticFiles/src/StaticFileContext.cs:297]: I think @Tratcher meant something like this in https://github.com/dotnet/aspnetcore/pull/45062#discussion_r1022179874:
```c#
_logger.Handled(statusCode, SubPath);
return ApplyResponseHeadersAsync(statusCode);
```

File: src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs [hunk 5]
```diff
@@ -201,6 +201,142 @@ private async Task FoundFile_Served(string baseUrl, string baseDir, string reque
...
+
+        using (var fileProvider = new PhysicalFileProvider(Path.Combine(AppContext.BaseDirectory, baseDir)))
+        {
+            using var host = await StaticFilesTestServer.Create(app => app.UseStaticFiles(new StaticFileOptions
+            {
+                RequestPath = new PathString(baseUrl),
+                FileProvider = fileProvider,
+                OnPrepareResponse = context =>
+                {
+                    onPrepareExecutionTime = DateTime.Now;
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:308]: don't rely on timestamps, use bools instead.
```suggestion
                    Assert.False(syncCallbackInvoked);
                    Assert.False(asyncCallbackInvoked);
                    syncCallbackInvoked = true;
```

File: src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs [hunk 6]
```diff
@@ -201,6 +201,142 @@ private async Task FoundFile_Served(string baseUrl, string baseDir, string reque
...
+                },
+                OnPrepareResponseAsync = context =>
+                {
+                    onPrepareAsyncExecutionTime = DateTime.Now;
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:312]:
```suggestion
                    Assert.True(syncCallbackInvoked);
                    Assert.False(asyncCallbackInvoked);
                    asyncCallbackInvoked = true;
```

File: src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs [hunk 7]
```diff
@@ -201,6 +201,142 @@ private async Task FoundFile_Served(string baseUrl, string baseDir, string reque
...
+            {
+                var fileContents = new byte[stream.Length];
+                stream.Read(fileContents, 0, (int)stream.Length);
+                Assert.True(responseContent.SequenceEqual(fileContents));
+            }
+        }
+        Assert.NotEqual(DateTime.MinValue, onPrepareExecutionTime);
+        Assert.NotEqual(DateTime.MinValue, onPrepareAsyncExecutionTime);
+
+        Assert.True(onPrepareExecutionTime.Ticks < onPrepareAsyncExecutionTime.Ticks);
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:337]: Assert.True(syncCallbackInvoked);
                    Assert.True(asyncCallbackInvoked);

File: src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs [hunk 8]
```diff
@@ -201,6 +201,142 @@ private async Task FoundFile_Served(string baseUrl, string baseDir, string reque
...
+    public async Task OnPrepareResponse_Executed_Test()
+    {
+        var baseUrl = "";
+        var baseDir = @".";
+        var requestUrl = "/testDocument.Txt";
+
+        var onPrepareResponseExecuted = false;
+
+        using (var fileProvider = new PhysicalFileProvider(Path.Combine(AppContext.BaseDirectory, baseDir)))
+        {
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:214]:
```suggestion
        using var fileProvider = new PhysicalFileProvider(Path.Combine(AppContext.BaseDirectory, baseDir));
```

**Review by Tratcher — APPROVED** [review]

File: src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs [hunk 9]
```diff
@@ -201,6 +201,142 @@ private async Task FoundFile_Served(string baseUrl, string baseDir, string reque
...
+            Assert.Equal(response.Content.Headers.ContentLength, responseContent.Length);
+            Assert.NotNull(response.Headers.ETag);
+
+            using (var stream = fileInfo.CreateReadStream())
+            {
+                var fileContents = new byte[stream.Length];
+                stream.Read(fileContents, 0, (int)stream.Length);
+                Assert.True(responseContent.SequenceEqual(fileContents));
+            }
+        }
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:240]: You still need to remove this brace and fix the formatting.

File: src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs [hunk 10]
```diff
@@ -293,48 +290,49 @@ public async Task OnPrepareResponse_Execution_Order_Test()
...
+        Assert.NotNull(response.Headers.ETag);
+
+        using (var stream = fileInfo.CreateReadStream())
+        {
+            var fileContents = new byte[stream.Length];
+            stream.Read(fileContents, 0, (int)stream.Length);
+            Assert.True(responseContent.SequenceEqual(fileContents));
         }
 
+
```
Comment (JamesNK) [inline src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:333]:
```suggestion
```

File: src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs [hunk 11]
```diff
@@ -201,6 +201,140 @@ private async Task FoundFile_Served(string baseUrl, string baseDir, string reque
...
+        Assert.NotNull(response.Headers.ETag);
+
+        using (var stream = fileInfo.CreateReadStream())
+        {
+            var fileContents = new byte[stream.Length];
+            stream.Read(fileContents, 0, (int)stream.Length);
+            Assert.True(responseContent.SequenceEqual(fileContents));
+        }
+
+
```
Comment (JamesNK) [inline src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:333]:
```suggestion
```

File: src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs [hunk 12]
```diff
@@ -201,6 +201,139 @@ private async Task FoundFile_Served(string baseUrl, string baseDir, string reque
         }
     }
 
+    [Fact]
+    public async Task OnPrepareResponse_Executed_Test()
+    {
+        var baseUrl = "";
+        var baseDir = @".";
+        var requestUrl = "/testDocument.Txt";
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:209]: Linux case sensitivity issue. The current version was copied from a Windows only test.
```suggestion
        var requestUrl = "/TestDocument.txt";
```

File: src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs [hunk 13]
```diff
@@ -201,6 +201,139 @@ private async Task FoundFile_Served(string baseUrl, string baseDir, string reque
...
+
+        Assert.True(onPrepareResponseExecuted);
+    }
+
+    [Fact]
+    public async Task OnPrepareResponseAsync_Executed_Test()
+    {
+        var baseUrl = "";
+        var baseDir = @".";
+        var requestUrl = "/testDocument.txt";
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:249]:
```suggestion
        var requestUrl = "/TestDocument.txt";
```

File: src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs [hunk 14]
```diff
@@ -201,6 +201,139 @@ private async Task FoundFile_Served(string baseUrl, string baseDir, string reque
...
+
+        Assert.True(onPrepareResponseExecuted);
+    }
+
+    [Fact]
+    public async Task OnPrepareResponse_Execution_Order_Test()
+    {
+        var baseUrl = "";
+        var baseDir = @".";
+        var requestUrl = "/testDocument.txt";
```
Comment (Tratcher) [inline src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:291]:
```suggestion
        var requestUrl = "/TestDocument.txt";
```

Comment (stefannikolei) [reply src/Middleware/StaticFiles/test/UnitTests/StaticFileMiddlewareTests.cs:291]: sorry for that much work...

**Pull request merged**